src/check.h

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_CHECK_H
+#define BITCOIN_CHECK_H
+
+#include <cstdio>
+#include <cstdlib>
+
+// DCHECK(expression)
+// ==================
+// DCHECK(...) is used to document assumptions.
+//
+// Bitcoin is always compiled with assertions enabled. assert(...)
+// is thus only appropriate for documenting critical assumptions
+// where a failed assumption should lead to immediate abortion also
+// in production environments.
+//
+// DCHECK(...) works like assert(...) if -DABORT_ON_FAILED_DCHECK
+// and is side-effect safe.
+//
+// More specifically:
+//
+// * If compiled with -DABORT_ON_FAILED_DCHECK (set by --enable-debug):
+//   Evaluate expression and abort if expression is falsy.
+//
+// * Otherwise:
+//   Evaluate expression and continue execution.
+//
+// ABORT_ON_FAILED_DCHECK should not be set in production environments.
+//
+// Example
+// =======
+// int main(void) {
+//     DCHECK(IsFoo());
+//     ...
+// }
+//
+// If !IsFoo() and -DABORT_ON_FAILED_DCHECK, then:
+//     filename.cpp:123: main: DCHECK(IsFoo()) failed.
+//     Aborted
+// Otherwise the execution continues.
+
+#ifdef ABORT_ON_FAILED_DCHECK
+#define DCHECK_FAILURE(expression, file, line, func)       \
+    do {                                                  \
+        fprintf(stderr, "%s:%d: %s: DCHECK(%s) failed.\n", \
+            file, line, func, expression);                \
+        std::abort();                                     \
+    } while (false)
+#else
+#define DCHECK_FAILURE(expression, file, line, func) ((void)0)
+#endif
+
+#define DCHECK(expression)                                             \
+    do {                                                              \
+        if (!(expression)) {                                          \
```

Comment:
It's possible I misunderstood part of the intent of this PR, but I don't think the expression should be evaluated in non-debug builds. Few reasons:

- It would encourage putting code that has external side effects in DCHECK macros, which would make it easier not to notice the effects of invoking code, and be bad for readability.
- It would encourage general misuse of the DCHECK macro. Like I wrote in my earlier comment I think the best use of DCHECK is just to state basic assumptions that code makes so it is more readable, and to help catch simple bugs during development. It shouldn't be used as lazy way to handle errors in release builds and avoid having to write readable error messages that put errors into context.
- It could potentially make release code less efficient.
- It would be inconsistent with other DCHECK implementations. Neither the glog nor folly DCHECK macros have side effects in release builds (see https://github.com/bitcoin/bitcoin/pull/16136#issuecomment-501779934) .

My suggestion would be to rename `ABORT_ON_FAILED_DCHECK` to `ENABLE_DCHECK` and when it's not defined either make `DCHECK` a no-op like folly, or make it compile the condition expression without executing it like glog.

Reply:
It seemed like the consensus in #4576 was to make the expression evaluated also in non-debug builds. FWIW, that is consistent with how it is done in `secp256k1`.

I don't have a strong opinion here: I'll let others chime in and adjust the code to meet consensus.

Reply:
Seems like clarifying the intent of the PR would be a good first step...

I think there's four things we could have:

 * `assert(chk)` -- always evaluate, always abort if it failes
 * `CHECK(chk)` -- always evaluate, but only abort if in regtest or similar; maybe log errors if not aborting.
 * `debug_assert(chk)` -- evaluate condition only when invoked with a special option like `-debugchecks=1`, which is always off by default, and can be disabled at compile time, allowing the code to dropped out. these checks could be slow and heavyweight. they probably wouldn't be run in travis.
 * `traditional_assert(chk)` -- evaluate for test builds, but don't evaluate in release builds

I think the `traditional_assert` style probably makes test and release builds different in ways that are hard to ensure are correct (in Haskell or Rust maybe we could ensure that chk is side effect free via the type-system, but that doesn't seem plausible in C++), so I don't think that's a good idea.

I could see any of the others being reasonable though.